scss/_patterns_table-hidden.scss

```diff
@@ -0,0 +1,34 @@
+@import 'settings';
+
+// mixin for hiding table columns
+@mixin vf-u-hide-table-column {
+  opacity: 0;
+  overflow: hidden; // make sure content doesn't overflow
+  padding: 0;
+  white-space: nowrap; // prevent content from wrapping in 0 width column
+  width: 0;
+}
+
+@mixin vf-p-table-hidden {
+  .p-table__cell--hidden {
+    @include vf-u-hide-table-column;
+
+    &--small {
```

Comment:
I don't think we want to start double modifying classes could this be `p-table__cell-hidden--small`?

Reply:
@anthonydillon True, I don't like double modifiers either, but `hidden` is modifier on `cell` element, so `p-table__cell--hidden-small`?

Reply:
Doesn't feel consistent with the rest of the framework. Considering the options:

| Interface | Usage |
| --- | --- |
| Table pattern element | `p-table-cell--hidden-small` |
| Table pattern utility | `u-hide-cell--small` |
| Extend the existing utility | `u-hide--small` |
| State class | `is-hidden--small` |

Please add our suggest and options that are not here. Looking at the usage table I believe extending the existing utility pattern feels like the most elegant approach.
 
**Foot note**
A part for the fact a dislike this approach of responsive table management. I believe the same content and functionally should be available on all screen sizes and devices. I have lost this battle a number of times.

Reply:
Sure, I agree with consistency, and consistency is why I started with `--hidden--small` (to keep `--small` as a modifier itself).

While we currently don't have `p-table` I believe in long term we will start to need that for more complex tables (modular table component). So I think we can start assuming that base table styling (HTML only) is for basic static tables, but any table functionality (icons, hiding, etc) would require `p-table` family of classes.

With that assumption to add icons to cells we added `p-table__cell--icon-placeholder`. `p-table` pattern, `__cell` child element `--icon-placeholder` modifier. 
Similarly, for consistency we would hide the cells using `p-table__cell--hidden`.  But the problem starts with modifiers of this hidden modifier, should they be `p-table__cell--hidden--small` (double modifier)? or `p-table__cell--hidden-small` (part of the same modifier, but inconsistent with rest of the framework).

In other places where we have multiple modifiers on same element we tend to use `is-something` kinds of class names. It's not BEM, and I'm not a huge fan of those, but they serve their purpose.
I don't like the idea of having just `is-hidden--small` (and scoping it to table HTML elements like td), but if we treat table as `p-table` pattern, and cell as `p-table__cell` we could have `p-table__cell.is-hidden--small`. So `is-hidden--small` state class, but only on a pattern element `p-table__cell`.

| Interface | Usage |
| --- | --- |
| State class on table pattern element | `p-table__cell.is-hidden--small` |

Reply:
@anthonydillon thoughts? ^

Reply:
Yeah I agree feels strange to create a new global sounding state class but it is consistent with the rest of the framework. What about using a different adverb. Maybe something like: `p-table__cell.is-removed--small` which would make it clearer and remove the confusion between this class and the `u-hide--small` class.

I keep coming back to the fact people familiar with Vanilla are likely to use `u-hide--small` before having to check why that doesn't work as expected. I wonder if we shouldn't just set `.p-table .u-hide` in the table pattern scss file. Then we don't need to document anything particularly differently.

Reply:
Thanks @anthonydillon, there is a lot of back and forth here, but valuable feedback to get us to good compromise.

> Yeah I agree feels strange to create a new global sounding state class but it is consistent with the rest of the framework. What about using a different adverb. Maybe something like: `p-table__cell.is-removed--small` which would make it clearer and remove the confusion between this class and the `u-hide--small` class.

We could try with different verb, but `removed` sounds more permanent than `hidden` which seems to be more accurate in case if responsively hiding columns. I thought about `collapsed` (which would be more in line with what the class actually does - collapses the column to 0 width without removing it) but this on the other hand sounds like something that can also be 'expanded'.

> I keep coming back to the fact people familiar with Vanilla are likely to use `u-hide--small` before having to check why that doesn't work as expected. I wonder if we shouldn't just set `.p-table .u-hide` in the table pattern scss file. Then we don't need to document anything particularly differently.

I agree that reusing `u-hide` and redefining it for tables is the least invasive approach. It doesn't introduce new class names, removes from developer the need to understand different concepts and keeps the complexity within the framework. But, this does mean there is a bit of complexity added.

First of all, it can't be just `.p-table .u-hide` - these new styles are specifically for table columns, not anything in the table, so we need `.p-table td.u-hide, .p-table th.u-hide`, and I guess people would expect it working consistently across the tables, not just with this `p-table` class, so why not `table td.u-hide, .table th.u-hide`, but this will again break in the expanding table which we want excluded from this...

The other thing is that it creates a dependency between a pattern and utility. And the question is where do we maintain this dependency. Should pattern define its exceptions to how utilities work, or should utility define how it works in different contexts? Either side is not perfect.
The other aspect of this is overriding and specificity. Utilities are built to override any other styling, they use !important all over the place. So it's even harder to override them - and that's what we would need to do to have `.p-table u-hide`. Sure, higher specificity and another bunch of !importants should do the trick, but I'm not sure if that should be our goto approach.

At this point, I am leaning towards implementing some variant of `.p-table th.u-hide, .p-table td.u-hide`, because, as you mentioned, it doesn't introduce new artifacts and we maybe don't need to document anything new (although `.p-table` would be kinda new, as it's not mentioned anywhere yet, because was never directly used). And this fix is also needed to unblock react-components for MAAS so would be good to release something soon (or we need to put a workaround in react-components).

Reply:
Yes, I like that solution with a slight twist. `th.u-hide, td.u-hide` which divorces the cross pattern/utility dependency. 

Regarding where it resides I believe we should add this variant to `_utilities_hide.scss`. The utility code has to live up to the promise that it will "hide" things as the user expects in all cases. Therefore, it feels right from a maintenance and responsibility point of view for this code to live there. 
 - **Maintenance**: because if there was an issue with `u-hide` not doing what I expected in any situation the first place I would go would be `_utilities_hide.scss`. It can also help us to rationale overrides but using `:not()` on the main utility and providing an extension very near each other in the code.
- **Responsibility**: I think of a utility as a promise to the user and therefore should handle any complexities that, that promise entails.

Reply:
@anthonydillon `th.u-hide, td.u-hide` is what I did in react components first, and this resulted in a broken expanding table. The problem with utility that addresses ALL tables is that we need to EXCLUDE p-table-expanding, because it uses flex layout instead of table layout and needs to use standard (display: none) u-hide.

If we say that `u-hide` should be responsible for how it works it means we need to do `td.u-hide, th.u-hide` override for tables but create another override for .p-expanding table. So (in simplified form):

```scss
.u-hide {
   display: none !important;
}

td.u-hide, th.u-hide {
   display: table-cell !important;
   width: 0 !important;
   // etc...
}

.p-table-expanding td.u-hide, .p-table-expanding th.u-hide {
   display: none !important;
}
```

Reply:
That seems reasonable as a temporary fix until 3.0 when we require the user to switch from using `u-hide` to using `aria-hidden='true'` in the expanding pattern.